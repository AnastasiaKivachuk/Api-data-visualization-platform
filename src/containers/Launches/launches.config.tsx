import * as React from 'react';

import {GridColDef, GridValueGetterParams} from '@mui/x-data-grid';

import {getDateWithFormat} from "../../helpers/functions.helpers";
import {DATE_FORMAT} from "../../constants/global.constants";
import {LAUNCH_ROUTE} from "../../constants/routes.constants";
import Button from "@mui/material/Button";
import Link from "next/link";

export const launchesColumns: GridColDef[] = [
    {field: 'id', headerName: 'id', sortable: false, width: 70},
    {
        field: 'mission_name', headerName: 'Mission Name', width: 200,
        valueGetter: (params: GridValueGetterParams) => params.row.mission_name || '-'
    },
    {
        field: 'launch_date_utc', headerName: 'Launch Date', width: 200,
        valueGetter: (params: GridValueGetterParams) => getDateWithFormat(params.row.launch_date_utc, DATE_FORMAT.DATE_WORD)
    },
    {
        field: 'rocket_name',
        headerName: 'Rocket Name',
        valueGetter: (params: GridValueGetterParams) => params.row.rocket.rocket_name || '-',
        width: 200
    },
    {
        field: 'site_name',
        headerName: 'Site',
        valueGetter: (params: GridValueGetterParams) => params.row.launch_site.site_name || '-',
        width: 200
    },
    {
        field: 'button',
        headerName: '',
        renderCell: (params: GridValueGetterParams) => <Link href={LAUNCH_ROUTE.replace('[id]', params.row.id)}>
            <Button>
                Show Details
            </Button></Link>,
        width: 200,
        sortable: false,
        type: 'actions',
    }
];