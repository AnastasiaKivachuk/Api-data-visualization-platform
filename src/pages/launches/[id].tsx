import type {NextPage} from 'next'
import React from "react";

import {getLaunch} from "../../services/api/launch.api";
import {LaunchWithDetailsInterface} from "../../interfaces/lauch.interface";
import Launch from "../../containers/Launch";
import Custom404Page from "../404";


export async function getServerSideProps(context: { query: { id: string; }; }) {
    try {
        const {data} = await getLaunch(String(context.query.id));
        return {
            props: {
                launch: data.launch,
            },
        };
    } catch (err) {
        return {props: {isError: true}};
    }
}

type Props = {
    launch?: LaunchWithDetailsInterface
    isError?: boolean
}

const LaunchPage: NextPage<Props> = ({launch, isError}) => {
    return (
        isError ? <Custom404Page/> : <Launch launch={launch}/>
    )
}

export default LaunchPage
