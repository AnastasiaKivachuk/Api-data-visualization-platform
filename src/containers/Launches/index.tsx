import React from "react";
import Typography from "@mui/material/Typography";
import {LaunchesInterface} from "../../interfaces/lauches.interface";

import {DataGrid} from "@mui/x-data-grid";
import {launchesColumns} from "./launches.config";

const Launches = ({launches}: LaunchesInterface) => {
    return (
        <>
            <Typography variant="h2" textAlign='center'>Launches</Typography>

            <div style={{height: 700, width: '100%'}}>
                {(launches || [])?.length > 0 ?
                    <DataGrid
                        rows={launches || []}
                        columns={launchesColumns}
                        pageSize={10}
                        rowsPerPageOptions={[10]}
                    /> :
                    <Typography variant="h4" textAlign='center'>No Data</Typography>
                }
            </div>
        </>
    )
}

Launches.defaultProps = {
    launches: []
}

export default Launches
