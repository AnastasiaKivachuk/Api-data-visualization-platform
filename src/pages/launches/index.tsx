import type { NextPage } from 'next'
import React from "react";

import Launches from "../../containers/Launches";
import {getLaunches} from "../../services/api/launches.api";
import {LaunchesInterface} from "../../interfaces/lauches.interface";
import Custom404Page from "../404";


export async function getStaticProps() {
  try {
    const {data} = await getLaunches();
    return {
      props: {
        launches: data.launches,
      },
    };
  } catch (err) {
    return {props: {}};
  }
}

interface Props extends LaunchesInterface {
  isError?: boolean
}

const LaunchesPage: NextPage<Props> = ({launches, isError}) => {
  return (
      isError ? <Custom404Page/> : <Launches launches={launches}/>
  )
}

export default LaunchesPage
