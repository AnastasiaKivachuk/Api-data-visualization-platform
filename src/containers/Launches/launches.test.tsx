import {shallow} from "enzyme";
import React from "react";

import Launches from "./index";
import {LaunchesInterface} from "../../interfaces/lauches.interface";
import {launchesColumns} from "./launches.config";

const MockLaunchesTable = [{
    id: "13",
    launch_date_utc: "2014-01-06T18:06:00.000Z",
    launch_site: {
        site_name: "CCAFS SLC 40"
    },
    mission_name: "Thaicom 6",
    rocket:
        {rocket_name: "Falcon 9"}
}]

const setUp = (props: LaunchesInterface | {}) => shallow(<Launches {...props}/>);

describe("Launches component", () => {

    it("should render Launches component with default props", () => {
        const component = setUp({});
        expect(component).toMatchSnapshot();
    })

    it("should render Launches component with props", () => {
        const component = setUp({launches: MockLaunchesTable});
        expect(component).toMatchSnapshot();
    })

    it("hide table and show info if no props", () => {
        const component = setUp({});
        const text = ['Launches', 'No Data'];
        expect(component.find('Memo(ForwardRef(DataGrid))').length).toBe(0)
        expect(component.find('ForwardRef(Typography)').length).toBe(2)
        component.find('ForwardRef(Typography)').map((item, index) => {
            expect(item.text()).toBe(text[index]);
        })
    })

    it("hide info and show table if props", () => {
        const component = setUp({launches: MockLaunchesTable});
        expect(component.find('Memo(ForwardRef(DataGrid))').length).toBe(1)
        expect(component.find('ForwardRef(Typography)').length).toBe(1)
        expect(component.find('ForwardRef(Typography)').text()).toBe('Launches')
    })

    it("check usersColumns config", () => {
        expect(launchesColumns).toMatchSnapshot();
    })
})
