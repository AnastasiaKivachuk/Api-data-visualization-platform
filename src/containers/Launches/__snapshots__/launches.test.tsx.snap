// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launches component check usersColumns config 1`] = `
Array [
  Object {
    "field": "id",
    "headerName": "id",
    "sortable": false,
    "width": 70,
  },
  Object {
    "field": "mission_name",
    "headerName": "Mission Name",
    "valueGetter": [Function],
    "width": 200,
  },
  Object {
    "field": "launch_date_utc",
    "headerName": "Launch Date",
    "valueGetter": [Function],
    "width": 200,
  },
  Object {
    "field": "rocket_name",
    "headerName": "Rocket Name",
    "valueGetter": [Function],
    "width": 200,
  },
  Object {
    "field": "site_name",
    "headerName": "Site",
    "valueGetter": [Function],
    "width": 200,
  },
  Object {
    "field": "button",
    "headerName": "",
    "renderCell": [Function],
    "sortable": false,
    "type": "actions",
    "width": 200,
  },
]
`;

exports[`Launches component should render Launches component with default props 1`] = `
<Fragment>
  <ForwardRef(Typography)
    textAlign="center"
    variant="h2"
  >
    Launches
  </ForwardRef(Typography)>
  <div
    style={
      Object {
        "height": 700,
        "width": "100%",
      }
    }
  >
    <ForwardRef(Typography)
      textAlign="center"
      variant="h4"
    >
      No Data
    </ForwardRef(Typography)>
  </div>
</Fragment>
`;

exports[`Launches component should render Launches component with props 1`] = `
<Fragment>
  <ForwardRef(Typography)
    textAlign="center"
    variant="h2"
  >
    Launches
  </ForwardRef(Typography)>
  <div
    style={
      Object {
        "height": 700,
        "width": "100%",
      }
    }
  >
    <Memo(ForwardRef(DataGrid))
      columns={
        Array [
          Object {
            "field": "id",
            "headerName": "id",
            "sortable": false,
            "width": 70,
          },
          Object {
            "field": "mission_name",
            "headerName": "Mission Name",
            "valueGetter": [Function],
            "width": 200,
          },
          Object {
            "field": "launch_date_utc",
            "headerName": "Launch Date",
            "valueGetter": [Function],
            "width": 200,
          },
          Object {
            "field": "rocket_name",
            "headerName": "Rocket Name",
            "valueGetter": [Function],
            "width": 200,
          },
          Object {
            "field": "site_name",
            "headerName": "Site",
            "valueGetter": [Function],
            "width": 200,
          },
          Object {
            "field": "button",
            "headerName": "",
            "renderCell": [Function],
            "sortable": false,
            "type": "actions",
            "width": 200,
          },
        ]
      }
      pageSize={10}
      rows={
        Array [
          Object {
            "id": "13",
            "launch_date_utc": "2014-01-06T18:06:00.000Z",
            "launch_site": Object {
              "site_name": "CCAFS SLC 40",
            },
            "mission_name": "Thaicom 6",
            "rocket": Object {
              "rocket_name": "Falcon 9",
            },
          },
        ]
      }
      rowsPerPageOptions={
        Array [
          10,
        ]
      }
    />
  </div>
</Fragment>
`;
